in-progress/7025-instrumenting-the-node-with-open-telemetry.md

```diff
@@ -0,0 +1,176 @@
+
+## Interface
+
+Enabling metrics in the node would require running a Prometheus instance to ingest all of this data and a Grafana dashboard to render the data. As part of this work the Docker Compose file used to run the Sandbox will be updated to optionally start these two services up.
+
+## Implementation
+
+OpenTelemetry supports three types of instrumentations: metrics, traces and logs. The immediate goal is to get structured metrics from the node. Traces and logs will be left for a future iteration.
+
+The OpenTelemetry framework is made up of two components: the API and the SDK. Library code is meant to import and use the API to emit traces, metrics and logs while applications are meant to initialize the SDK. The SDK acts as a "backend" for the API, without it every trace, metric and log become no-ops. This would allow to conditionally initialize the SDK depending on whether stats have been enabled by the user.
+
+> [!NOTE]
+> The `@opentelemetry/api` and `@opentelemetry/sdk-node` packages use global scope for the backend. This means we can't have two services initialize two SDKs in the same NodeJS process as they'd clash with each other (e.g. running both the node and the pxe in the same process and both initialize an SDK instance.)
+
+### Naming
+
+The OpenTelemetry specification already includes guidelines on telemetry naming. We'll follow the established guidelines:
+
+- we will prefix all metrics with `aztec`
+- we will use full stops `.` to separate components of a name (e.g. `aztec.circuit.simulation`)
+- we will use base units for values (e.g. `bytes` over `kilobytes` and `seconds` over `milliseconds`)
+- we will _not_ include the unit in the name (there's a separate attribute for that, see code examples)
+
+TBD `meter` naming conventions (ie. `metrics.getMeter('RealPublicKernelCircuitSimulator')`)
+
+[Metrics guidelines & naming conventions](https://opentelemetry.io/docs/specs/semconv/general/metrics/)
+
+> [!NOTE]
+> Prometheus does not accept full-stops `.` in metric names and will replace them with underscores `_` currently. This will change once it fully supports the OpenTelemetry specification. See [prometheus/prometheus#13095](https://github.com/prometheus/prometheus/issues/13095)
+
+### Benchmark metrics
+
+All call sites that emit benchmark stats would be extended to also emit an identical metric via OTel:
+
+```ts
+// yarn-project/simulator/src/public/public_kernel.ts
+import { metrics } from '@opentelemetry/api';
+export class RealPublicKernelCircuitSimulator implements PublicKernelCircuitSimulator {
+  // ...
+  private meter = metrics.getMeter('RealPublicKernelCircuitSimulator');
+  private circuitSimulationMetric = this.meter.createHistogram('aztec.circuit.simulation.duration', {
+    unit: 's',
+  });
+
+  public async publicKernelCircuitSetup(input: PublicKernelCircuitPrivateInputs): Promise<PublicKernelCircuitPublicInputs> {
+    // ...
+
+    this.circuitSimulationMetric.record(duration / 1000, {
```

Comment:
`duration * 1000`

Reply:
Sorry, forgot to get back to you on this. The latest recommendation is to use seconds over milliseconds for duration metrics. I've updated the proving code to return a `durationMs` field instead of the unitless `duration` we had before.

https://github.com/open-telemetry/opentelemetry-specification/issues/2977